### PR TITLE
core/web: /health - and support for HTML & Plaintext

### PR DESCRIPTION
**Pull request by jmank88**
The shape of our internal service architecture is not very visible. We have a fully qualified service id that we use in our logs and our health metrics, but it is still difficult to get a high level view of what is going on, or even just running.
We have a `/health` endpoint that returns compact JSON. This PR extends support for returning HTML or plaintext as well.

Screenshot from 2023-12-12 17-02-52

image

Requires:
- #11558

### CODE REVIEW

File: core/web/health_controller.go [hunk 1]
```diff
@@ -93,7 +99,200 @@ func (hc *HealthController) Health(c *gin.Context) {
...
+		}
+		return
+	}
 
 	// return a json description of all the checks
-	jsonAPIResponse(c, checks, "checks")
+	jsonAPIResponseWithStatus(c, checks, "checks", status)
+}
+
+func writeTextTo(w io.Writer, checks []presenters.Check) error {
```
Comment (krehermann) [inline core/web/health_controller.go:133]: why is all this extra new stuff in this change? is it needed to fix the evm healthreport? it seems like a feature that is orthogonal to the health fix. what is the feature getting us relative to what's already implemented?

Comment (jmank88) [reply core/web/health_controller.go:133]: This was the feature that exposed the problem. I can separate the fix, but we don't have good existing tests, so I'd have to write an intermediate test that worked differently.

File: core/web/health_controller.go [hunk 2]
```diff
@@ -93,7 +99,200 @@ func (hc *HealthController) Health(c *gin.Context) {
...
+
+	Subs checkTree
+}
+
+type checkTree map[string]checkNode
+
+func newCheckTree(checks []presenters.Check) checkTree {
+	slices.SortFunc(checks, func(a, b presenters.Check) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
```
Comment (jmank88) [inline core/web/health_controller.go:173]: I can drop this now that we always sort...unless that is too costly for the default JSON case?
```suggestion
```

File: core/chains/evm/gas/models.go [hunk 3]
```diff
@@ -71,19 +71,31 @@ func NewEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, ge
...
+			return NewSuggestedPriceEstimator(lggr, ethClient)
+		}
 	default:
 		lggr.Warnf("GasEstimator: unrecognised mode '%s', falling back to FixedPriceEstimator", s)
-		return NewWrappedEvmEstimator(NewFixedPriceEstimator(geCfg, bh, lggr), df, l1Oracle)
+		est = func(l logger.Logger) EvmEstimator {
+			return NewFixedPriceEstimator(geCfg, bh, lggr)
+		}
 	}
+	return NewWrappedEvmEstimator(lggr, est, df, l1Oracle)
```
Comment (patrickhuie19) [inline core/chains/evm/gas/models.go:98]: What do you think of a factory pattern?

Interface:
```
type EstimatorFactory interface {
    CreateEstimator(logger.Logger, evmclient.Client, Config, evmconfig.GasEstimator, rollups.L1Oracle) EvmEstimator
}
```

Example Implementations
```
// ArbitrumEstimatorFactory creates an ArbitrumEstimator
type ArbitrumEstimatorFactory struct{}

func (f *ArbitrumEstimatorFactory) CreateEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, geCfg evmconfig.GasEstimator, l1Oracle rollups.L1Oracle) EvmEstimator {
    return NewArbitrumEstimator(lggr, geCfg, ethClient, ethClient)
}

// BlockHistoryEstimatorFactory creates a BlockHistoryEstimator
type BlockHistoryEstimatorFactory struct{}

func (f *BlockHistoryEstimatorFactory) CreateEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, geCfg evmconfig.GasEstimator, l1Oracle rollups.L1Oracle) EvmEstimator {
    bh := geCfg.BlockHistory()
    return NewBlockHistoryEstimator(lggr, ethClient, cfg, geCfg, bh, *ethClient.ConfiguredChainID())
}

// ... 
```
Switch case refactored: 
```
    var factory EstimatorFactory
    switch s {
    case "Arbitrum":
        factory = &ArbitrumEstimatorFactory{}
    case "BlockHistory":
        factory = &BlockHistoryEstimatorFactory{}
    // ... other cases
    default:
        lggr.Warnf("GasEstimator: unrecognised mode '%s', falling back to FixedPriceEstimator", s)
        factory = &FixedPriceEstimatorFactory{}
    }

    est := factory.CreateEstimator(lggr, ethClient, cfg, geCfg, l1Oracle)
    return NewWrappedEvmEstimator(lggr, est, df, l1Oracle)
```

Comment (jmank88) [reply core/chains/evm/gas/models.go:98]: An interface seems like overkill since there is no state to track in the factories. But regardless, the crux of the problem here is the service name derived from the logger. The code was acting like the wrapper owned the estimator, but the service name did not reflect that. This change was meant to correct that by moving estimator construction inside of the wrapper. We could pass the factory in, but then we have to pipe along extra parameters through the wrapper.

File: core/web/health_controller.go (hunk 1 above)
Comment (jmank88) [reply core/web/health_controller.go:133]: I split out the fix as #11558

File: core/chains/evm/gas/models.go (hunk 3 above)
Comment (patrickhuie19) [reply core/chains/evm/gas/models.go]: > We could pass the factory in, but then we have to pipe along extra parameters through the wrapper.

Could you give an example? I'm not seeing where the extra parameters would come from.

Comment (jmank88) [reply core/chains/evm/gas/models.go]: Moved to #11558 and named better there.

File: core/web/health_controller.go [hunk 4]
```diff
@@ -96,13 +99,202 @@ func (hc *HealthController) Health(c *gin.Context) {
...
+	case gin.MIMEHTML:
+		if err := newCheckTree(checks).WriteHTMLTo(c.Writer); err != nil {
+			hc.App.GetLogger().Errorw("Failed to write HTML health report", "err", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+		}
+		return
 
-	if testing.Testing() {
+	case gin.MIMEPlain:
 		slices.SortFunc(checks, func(a, b presenters.Check) int {
```
Comment (patrickhuie19) [inline core/web/health_controller.go:118]: Should this sort be removed since you sort before the switch statement now?

File: core/web/health_controller.go [hunk 5]
```diff
@@ -96,13 +99,202 @@ func (hc *HealthController) Health(c *gin.Context) {
...
+	Status string
+	Output string
+
+	Subs checkTree
+}
+
+type checkTree map[string]checkNode
+
+func newCheckTree(checks []presenters.Check) checkTree {
+	if !slices.IsSortedFunc(checks, func(a, b presenters.Check) int {
```
Comment (patrickhuie19) [inline core/web/health_controller.go:171]: why assert that the caller has to sort `checks` beforehand? Could we just sort here?

File: core/web/health_controller.go (hunk 4 above)
Comment (jmank88) [reply core/web/health_controller.go:118]: yeah

File: core/web/health_controller.go (hunk 5 above)
Comment (jmank88) [reply core/web/health_controller.go:171]: Because we sort for the other cases and I wanted to avoid double sorting and not leave things too brittle. We could just sort in the three separate places instead. That is probably simpler. Can make it a one-liner too, instead of three.

Comment (jmank88) [reply core/web/health_controller.go]: Updated these to: `slices.SortFunc(checks, presenters.CmpCheckName)`

**Review by samsondav — APPROVED** [review]
nice idea
